pkg/controller/issueapi/logic.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package issueapi
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/google/exposure-notifications-server/pkg/timeutils"
+	"github.com/google/exposure-notifications-verification-server/internal/project"
+	"github.com/google/exposure-notifications-verification-server/pkg/api"
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+	"github.com/google/exposure-notifications-verification-server/pkg/otp"
+	"github.com/google/exposure-notifications-verification-server/pkg/sms"
+	"go.opencensus.io/stats"
+)
+
+func (c *Controller) issue(ctx context.Context, request *api.IssueCodeRequest,
+	realm *database.Realm, user *database.User, authApp *database.AuthorizedApp) (*issueResult, *api.IssueCodeResponse) {
+	logger := logging.FromContext(ctx).Named("issueapi.issue")
+	var err error
+
+	// If this realm requires a date but no date was specified, return an error.
+	if realm.RequireDate && request.SymptomDate == "" && request.TestDate == "" {
+		return &issueResult{
+			obsBlame:    observability.BlameClient,
+			obsResult:   observability.ResultError("MISSING_REQUIRED_FIELDS"),
+			httpCode:    http.StatusBadRequest,
+			errorReturn: api.Errorf("missing either test or symptom date").WithCode(api.ErrMissingDate),
+		}, nil
+	}
+
+	// Validate that the request with the provided test type is valid for this realm.
+	if !realm.ValidTestType(request.TestType) {
+		return &issueResult{
+			obsBlame:    observability.BlameClient,
+			obsResult:   observability.ResultError("UNSUPPORTED_TEST_TYPE"),
+			httpCode:    http.StatusBadRequest,
+			errorReturn: api.Errorf("unsupported test type: %v", request.TestType),
```

Comment:
.WithCode?

Reply:
done. Also added a `quota_exceeded` code which seems to be missing